### ISSUE TITLE
Function parameter names are incorrectly inferred from variadiac tuples

### ISSUE THREAD

**Issue by Nvveen** [microsoft/TypeScript#40452]
**TypeScript Version:**  4.1.0-insiders.20200909

**Search Terms:**
typescript variadic tuple names curry argument
**Code**

```ts
type Curry<A extends unknown[], R> = A extends [infer Head, ...infer Tail] ? (...args: [Head]) => Curry<Tail, R> : R;
declare const curry: <Args extends unknown[], R>(fn: (...args: Args) => R) => Curry<Args, R>;
declare const f: (x: number, y: string, z: boolean) => string;
const g  = curry(f); // incorrectly inferred as (arg_0: number) => (arg_0: string) => (arg_0: boolean) => string;
```

**Expected behavior:**
The inferred type should have the correct argument names from `f`, like `(x: number) => (y: string) => (z: boolean) => string;`
**Actual behavior:**
`(arg_0: number) => (arg_0: string) => (arg_0: boolean) => string;`
**Playground Link:** 
https://tsplay.dev/nWPZJN

**Comment by RyanCavanaugh** [comment]
Parameter names are not guaranteed to be preserved through all transforms - you should think of this a convenience, not a correctness - but this might be a case we could make work

**Comment by Nvveen** [comment]
FYI, explicitly passing a tuple, even a tuple with labels, shows the same behaviour as above.

**Comment by aminpaks** [comment]
@RyanCavanaugh I think it's fair to ask for simple transformation to pass on the labels as in example below:
```ts
type Curry<A extends unknown[], R> = A extends [infer Head, ...infer Tail] ? (...args: [Head]) => Curry<Tail, R> : R;

type T0 = [x: number, y: string, z: boolean]
type T1 = Curry<T0, string>;
```

[Playground link](
https://www.typescriptlang.org/play?jsx=0&ts=4.0.2#code/C4TwDgpgBAwgrgJwSAPAQShAHsCA7AEwGco48BrPAewHc8BtAXQBooAlAPigF4oNtchEvQCWeAGYQEUABIQAhgVYA6VWMnSAKvJEAbRlAD8UABSrl8hAHMiALij05ixgEoeXeElTa9rTlHs2AG4AKBDQSChNAAYeByx7PDgAWwAjKVYQeyJgBDErVgAve1SqKl0FPEZw8GhNAEY4z2QUGNYcvLwrDlCQgHo+qABRLEgAY2B8qMbgKih0+xMEqCS0qTduLhMsqA78ja3i+bKK+TwD3dz8kKA)

**Comment by ahejlsberg** [comment]
The issue here is that only tuple types can carry along parameter names. In the example, though it is inferred from a tuple type, `Head` itself is not a tuple type, so any parameter name associated with the type is lost. It's not a trivial matter to preserve the name, though I agree it would be nice.

**Comment by aminpaks** [comment]
@ahejlsberg Right. Is it possible to ask if the `infer` is a tuple element to somewhat infer the label too?

**Comment by ahejlsberg** [comment]
Only way I can think of is to support inferring to consecutive variadic elements:

```ts
type Curry<A extends unknown[], R> = A extends [...infer Head, ...infer Tail] ? (...args: Head) => Curry<Tail, R> : R;
```

Currently we make no inferences in such cases because it isn't clear where to separate the source tuple, but we could consider a convention of inferring a single element tuple to the first position and a tuple with the remaining elements to the second position. This would be similar to what we do in template literal type inference.

**Comment by aminpaks** [comment]
What can be the different results of `A extends [...infer Head, ...infer Tail] ? ...` ?
I'm trying to understand if other options than making the first element `Head` a tuple would make more sense.

Also I'm so curious to know if there is a better way to infer the first element's label of a tuple. For example something like this `A extends [infer HeadLabel: Head, ...infer Tail] ? ...`

**Comment by brettmitchelldev** [comment]
I believe that this is not necessarily linked to inferring a tuple member; I recently experienced the same behavior when trying to constrain variadic function parameters while retaining parameter names without destructuring the tuple in any way. Here is a minimal example (note: the minimal example could totally be implemented in a simpler way, but my use case requires this strategy):

```typescript
type ConstrainedNamedVariadic <Constraint extends any[]> = <Args extends Constraint = Constraint> (...args: Args) => Args;
type ConstrainedFn = ConstrainedNamedVariadic <[p1: number, p2: string]>;
const constrainedFn: ConstrainedFn;
constrainedFn ();
// => const constrainedFn: <[p1: number, p2: string]>(p1: number, p2: string) => [p1: number, p2: string]
constrainedFn (123, 'str');
// => const constrainedFn: <[123, "str"]>(args_0: 123, args_1: "str") => [123, "str"]
```

I do recognize that this may be viewed as expected behavior from a compiler-internal perspective, but it would be nice to have the compiler preserve labels when they are not specified on the inferred type parameter.
